---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Methods

Raw data were processed using cellranger 6.0.2.

The sample CBF (non hashed) has been sequenced two times due to a low cell recovery on first sequencing.

Cell hashing for CBF and FLT3 sample was sequenced two times.

Fixed sample P8 was sequenced two times due to low cell recovery.

For runs sequenced two times, cellranger count was parametered in order to take two input run  [into account (see end of page)](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count).

Single-cell data are processed using the Seurat R package (Seurat version 4.0.4, R version 4.1.1)

The *Read10X* function from Seurat can read *cellranger count* output and convert it into a matrix.

```{r loading_data}
UPN23 <- Read10X(data.dir = "../../../out/cellranger/count_--sample_FLT3_R252_GRCh38-2020-A/cellranger/mkfastq/gpfs/tgml/reads/bcl/Run_252_190329_NS500637_0159_AHYGNKBGX9/Run_252/outs/filtered_feature_bc_matrix/")
UPN29 <- Read10X(data.dir = "../../../out/cellranger/count_CBF_sample_R252_255/outs/filtered_feature_bc_matrix/")
UPN23_UPN29_cellhashing <- Read10X(data.dir = "../../../out/cellranger/count_mRNA_R306_R309/count_mRNA_R306_R309//outs/filtered_feature_bc_matrix/")
```

The *CreateSeuratObject* function convert the matrix into a Seurat object that will store analysis. Two parameters are set when creating Seurat object:

* min.cells = 3 (default) -> Include features (genes) detected in at least this many cells.
* min.features = 200 (default) -> Include cells where at least this many features (genes) are detected.

```{r seurat_object_for_non_hashed}
Seurat_UPN23 <- CreateSeuratObject(counts = UPN23, project = "UPN23", min.cells = 0, min.features = 0)
Seurat_UPN29 <- CreateSeuratObject(counts = UPN29, project = "UPN29", min.cells = 0, min.features = 0)
Seurat_UPN23_UPN29_cellhashing <- CreateSeuratObject(counts = UPN23_UPN29_cellhashing, project = "UPN23_UPN29_cellhashing", min.cells = 0, min.features = 0)

# rm(UPN23, UPN29, UPN23_UPN29_cellhashing)
```

HTO data are processed using the [cite-seq count, v1.4.3](https://hoohm.github.io/CITE-seq-Count/) tool. Results are read using the *Read10X* from Seurat.

Only common cell barcode originating from HTO (from cite-seq count) and mRNA (from cellranger) data are kept. Those barcodes represent cells detected by cellranger and for which an HTO has been attributed.

```{r read_citeseq_count_information}
# Read umi_count tables for each sample
HTO_UPN23_UPN29_cellhashing <-  Read10X(data.dir = "../../../out/cite-seq_count/Run_306_Run_309/umi_count/", gene.column = 1)
# Remove Hashtag barcode from HTO rownames
rownames(HTO_UPN23_UPN29_cellhashing) <- gsub(pattern = "-.*", replacement = "", x = rownames(HTO_UPN23_UPN29_cellhashing))
# Add a "-1" at the end of cell barcodes for HTO data to have similar names as mRNA data
colnames(HTO_UPN23_UPN29_cellhashing) <- paste0(colnames(HTO_UPN23_UPN29_cellhashing),"-1")
# Filter cells that are detected in both mRNA and HTO
joint.bcs_UPN23_UPN29_cellhashing <- intersect(colnames(Seurat_UPN23_UPN29_cellhashing), colnames(HTO_UPN23_UPN29_cellhashing))
# Subset mRNA seurat object to shared cell barcodes
Seurat_UPN23_UPN29_cellhashing <- subset(x = Seurat_UPN23_UPN29_cellhashing, cells = joint.bcs_UPN23_UPN29_cellhashing)
```

Finally, the HTO matrices are normalized using the Centered Log-Ratio method (CLR). This method is recommended by Seurat developers. Then, the *MULTIseqDemux* function is used to assign HTO name to cells. For this function, the parameter *autoThresh = TRUE* was used. This parameter enable the function to perform automated threshold finding to define the best quantile.

```{r HTO_demultiplexing}
HTO_UPN23_UPN29_cellhashing_joints_bcs <- HTO_UPN23_UPN29_cellhashing[grep("HTO", rownames(HTO_UPN23_UPN29_cellhashing)), colnames(Seurat_UPN23_UPN29_cellhashing)]
rownames(HTO_UPN23_UPN29_cellhashing_joints_bcs) <- c("UPN29-cellhashing", "UPN23-cellhashing")
Seurat_UPN23_UPN29_cellhashing[["HTO"]] <- CreateAssayObject(counts = HTO_UPN23_UPN29_cellhashing_joints_bcs)
# Normalize HTO data, here we use centered log-ratio (CLR) transformation
Seurat_UPN23_UPN29_cellhashing <- NormalizeData(Seurat_UPN23_UPN29_cellhashing, assay = "HTO", normalization.method = "CLR", margin = 1)
# Perform HTO demultiplexing
Seurat_UPN23_UPN29_cellhashing <- MULTIseqDemux(Seurat_UPN23_UPN29_cellhashing, assay = "HTO", autoThresh = TRUE)

# Subset Seurat_UPN23_UPN29_cellhashing for the 2 HTO
Seurat_UPN23_UPN29_cellhashing_UPN23 <- subset(Seurat_UPN23_UPN29_cellhashing, cells = colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN23-cellhashing"])
Seurat_UPN23_UPN29_cellhashing_UPN29 <- subset(Seurat_UPN23_UPN29_cellhashing, cells = colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN29-cellhashing"])
```

```{r}
# Write HTO assignation to csv files
write.csv(Seurat_UPN23_UPN29_cellhashing@meta.data[,c("orig.ident", "MULTI_ID")], file = "../../../out/csv/HTO_UPN23_UPN29.csv")
write(colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN23-cellhashing"], 
file = "../../../out/csv/UPN23_cellbarcodes.csv")
write(colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN29-cellhashing"], 
file = "../../../out/csv/UPN29_cellbarcodes.csv")
```

<!-- The read number for the cellhashing is 871 026 117, in order to have the mean read per cell, we have to divide this number by the number of cells attributed to each HTO. We estimate the number of read of each sample in cellhashing data by getting the percentage of cell attributed to each HTO in the total sample and divie the total read number by those percentages. -->

To get the read number for cellhashing data, we exported the cellbarcode associated to each HTO, use the [subset-bam](https://github.com/10XGenomics/subset-bam) tool to extract mapped read corresponding to each sample. Then, mapped read for each cell were counted using a bash commands:

> grep CB:Z: UPN29.sam | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > UPN29_reads_per_barcode.txt                                  
> grep CB:Z: UPN23.sam | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > UPN23_reads_per_barcode.txt                                  

```{r}
UPN23_read_number <- read.delim("../../../out/subset-bam/UPN23_reads_per_barcode_nospaces.txt", header = F, sep = "\t")
UPN29_read_number <- read.delim("../../../out/subset-bam/UPN29_reads_per_barcode_nospaces.txt", header = F, sep = "\t")
# HTO_counts <- Read10X("../../../out/cite-seq_count/Run_306_Run_309/read_count", gene.column = 1)
# colnames(HTO_counts) <- paste0(colnames(HTO_counts), "-1")

Seurat_object_list <- list(Seurat_UPN23, Seurat_UPN29, Seurat_UPN23_UPN29_cellhashing_UPN23, Seurat_UPN23_UPN29_cellhashing_UPN29)
# 
# cellhashing_total_cell <- ncol(Seurat_UPN23_UPN29_cellhashing)
# percent_UPN23 <- ncol(Seurat_UPN23_UPN29_cellhashing_UPN23)/cellhashing_total_cell
# percent_UPN29 <- ncol(Seurat_UPN23_UPN29_cellhashing_UPN29)/cellhashing_total_cell
# UPN23_cellhashing_nreads <- percent_UPN23*871026117
# UPN29_cellhashing_nreads <- percent_UPN29*871026117
# 
mean_reads_per_cell <- c(171943491/ncol(Seurat_object_list[[1]]),
                         711710869/ncol(Seurat_object_list[[2]]),
                         mean(UPN23_read_number$V1), mean(UPN29_read_number$V1))
mean_reads_per_cell <- sapply(mean_reads_per_cell, format, digit = 3)
# 
table_colnames <- c("Estimated number of cells", "Total read number", "Mean reads per cell", "Median genes per cell")
# "Sequencing saturation", "Read mapped confidently to genome", "Fraction reads in cells", "Total genes detected", "Median UMI counts per cell"
Result_table <- data.frame(matrix(nrow = 4, ncol = 4), row.names = c("UPN23", "UPN29", "UPN23_cellhashing", "UPN29_cellhashing"))
colnames(Result_table) <- table_colnames
Result_table$`Estimated number of cells` <- sapply(Seurat_object_list, ncol)
Result_table$`Total read number` <- c(171943491, 711710869, sum(UPN23_read_number$V1), sum(UPN29_read_number$V1))
Result_table$`Mean reads per cell` <- mean_reads_per_cell
Result_table$`Median genes per cell` <- sapply(Seurat_object_list, function(x){median(x$nFeature_RNA)})
```


```{r}
#Clean memory
rm(HTO_UPN23_UPN29_cellhashing, HTO_UPN23_UPN29_cellhashing_joints_bcs,
   joint.bcs_UPN23_UPN29_cellhashing)

gc()
```

```{r}
# Create seurat object for non multiplexed data.
Seurat_UPN23 <- CreateSeuratObject(counts = UPN23, project = "UPN23", min.cells = 3, min.features = 200)
Seurat_UPN29 <- CreateSeuratObject(counts = UPN29, project = "UPN29", min.cells = 3, min.features = 200)
Seurat_UPN23_UPN29_cellhashing <- CreateSeuratObject(counts = UPN23_UPN29_cellhashing, project = "UPN23_UPN29_cellhashing", min.cells = 3, min.features = 200)

# Read umi_count tables for each sample
HTO_UPN23_UPN29_cellhashing <-  Read10X(data.dir = "../../../out/cite-seq_count/Run_306_Run_309/umi_count/", gene.column = 1)
# Remove Hashtag barcode from HTO rownames
rownames(HTO_UPN23_UPN29_cellhashing) <- gsub(pattern = "-.*", replacement = "", x = rownames(HTO_UPN23_UPN29_cellhashing))
# Add a "-1" at the end of cell barcodes for HTO data to have similar names as mRNA data
colnames(HTO_UPN23_UPN29_cellhashing) <- paste0(colnames(HTO_UPN23_UPN29_cellhashing),"-1")
# Filter cells that are detected in both mRNA and HTO
joint.bcs_UPN23_UPN29_cellhashing <- intersect(colnames(Seurat_UPN23_UPN29_cellhashing), colnames(HTO_UPN23_UPN29_cellhashing))
# Subset mRNA seurat object to shared cell barcodes
Seurat_UPN23_UPN29_cellhashing <- subset(x = Seurat_UPN23_UPN29_cellhashing, cells = joint.bcs_UPN23_UPN29_cellhashing)

HTO_UPN23_UPN29_cellhashing_joints_bcs <- HTO_UPN23_UPN29_cellhashing[grep("HTO", rownames(HTO_UPN23_UPN29_cellhashing)), colnames(Seurat_UPN23_UPN29_cellhashing)]
rownames(HTO_UPN23_UPN29_cellhashing_joints_bcs) <- c("UPN29-cellhashing", "UPN23-cellhashing")
Seurat_UPN23_UPN29_cellhashing[["HTO"]] <- CreateAssayObject(counts = HTO_UPN23_UPN29_cellhashing_joints_bcs)
# Normalize HTO data, here we use centered log-ratio (CLR) transformation
Seurat_UPN23_UPN29_cellhashing <- NormalizeData(Seurat_UPN23_UPN29_cellhashing, assay = "HTO", normalization.method = "CLR", margin = 1)
# Perform HTO demultiplexing
Seurat_UPN23_UPN29_cellhashing <- MULTIseqDemux(Seurat_UPN23_UPN29_cellhashing, assay = "HTO", autoThresh = TRUE)

# Subset Seurat_UPN23_UPN29_cellhashing for the 2 HTO
Seurat_UPN23_UPN29_cellhashing_UPN23 <- subset(Seurat_UPN23_UPN29_cellhashing, cells = colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN23-cellhashing"])
Seurat_UPN23_UPN29_cellhashing_UPN29 <- subset(Seurat_UPN23_UPN29_cellhashing, cells = colnames(Seurat_UPN23_UPN29_cellhashing)[Seurat_UPN23_UPN29_cellhashing$MULTI_ID == "UPN29-cellhashing"])

rm(HTO_UPN23_UPN29_cellhashing, HTO_UPN23_UPN29_cellhashing_joints_bcs, 
   joint.bcs_UPN23_UPN29_cellhashing)

gc()
```


